mlaut/benchmarking/base.py

```diff
@@ -0,0 +1,179 @@
+        raise NotImplementedError()
+
+    def _append_key(self, strategy_name, dataset_name):
+        """Append names of datasets and strategies to results objects during orchestration"""
+        if strategy_name not in self.strategy_names:
+            self.strategy_names.append(strategy_name)
+
+        if dataset_name not in self.dataset_names:
+            self.dataset_names.append(dataset_name)
+
+    def _generate_key(self, strategy_name, dataset_name, cv_fold, train_or_test):
+        raise NotImplementedError()
+
+    def __repr__(self):
+        class_name = self.__class__.__name__
+        return f"{class_name}(strategies={self.strategy_names}, datasets={self.dataset_names}, " \
+               f"cv_folds={self.cv.get_n_splits()})"
+
+    def save(self):
+        """Save results object as master file"""
+        NotImplementedError()
+
+    def _iter(self):
+        """Iterate over registry of results object"""
+        for strategy_name in self.strategy_names:
+            for dataset_name in self.dataset_names:
+                yield strategy_name, dataset_name
+
+
+class HDDBaseResults(BaseResults):
+
+    def __init__(self, path):
+        # validate paths
+        self._validate_path(path)
+
+        # set path
+        self._path = path
+
+        super(HDDBaseResults, self).__init__()
+
+    @property
+    def path(self):
+        return self._path
+
+    def save(self):
+        """Save results object as master file"""
+        file = os.path.join(self.path, "results.pickle")
+
+        # if file does not exist already, create a new one
+        if not os.path.isfile(file):
+            dump(self, file)
+
+        # if file already exists, update file adding new datasets, strategies and/or cv_folds
+        else:
+            results = load(file)
+            self.strategy_names = list(set(self.strategy_names + results.strategy_names))
+            self.dataset_names = list(set(self.dataset_names + results.dataset_names))
+            dump(self, file)
+
+    @staticmethod
+    def _validate_path(path):
+        """Helper function to validate paths"""
+        # check if path already exists
+        if os.path.exists(path):
+            if not os.path.isdir(path):
+                raise ValueError("path already exists and is not a directory")
+
+            elif os.path.isfile(os.path.join(path, "results.pickle")):
+                warn(f"Existing results file found in given path: {path}. "
+                     f"Results file will be updated")
+
+            elif len([file for file in os.listdir(path) if not file.startswith(".")]) > 0:
+                warn("path already exists and is not empty")
+
+
+class _PredictionsWrapper:
+    """Single result class to ensure consistency for return object when loading results"""
+
+    def __init__(self, strategy_name, dataset_name, index, y_true, y_pred, y_proba=None):
+        # check input format
+        if not all(isinstance(array, np.ndarray) for array in [y_true, y_pred]):
```

Comment:
probably cleaner to refactor this into separate methods, e.g. `check_y` and `check_names`, they will also be used to check the input of the orchestrator